### PR TITLE
Add reflection support for mobs, reflective vest, shuttle walls

### PR DESCRIPTION
**Pull request by graevy**
## About the PR

it addresses [this issue](https://github.com/space-wizards/space-station-14/issues/16095) by adding `ReflectComponent` to `BaseMobOrganic` and some mentioned items*. it contains some `SharedReflectSystem` logic for extensible handling of multiple equipment reflection layers. this is my first time with this repo/C#/ECS programming; I'm hesitant to refactor `ReflectComponent` (having both `Enabled` and `Layers` is clunky) before review.

*i don't personally remember shuttle walls or the syndie CE jumpsuit having reflection, but that was like 10 years ago.

**Media**

urist's revenge

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl:
- add: Added reflection to the reflective vest and shuttle walls

### CODE REVIEW

**Comment by keronshb** [comment]
Based. Will review today.

**Comment by Flareguy** [comment]
If you're adding reflection to the syndicate CE jumpsuit also add it to the Nanotrasen ship vs. ship variant please, or at least give it another similarly good buff. It'd be dumb if the NT-sided CE was just worse for no reason.

File: Resources/Prototypes/Entities/Clothing/OuterClothing/armor.yml [hunk 1]
```diff
@@ -94,6 +94,8 @@
         Slash: 0.9
         Piercing: 0.9
         Heat: 0.4 # this technically means it protects against fires pretty well? -heat is just for lasers and stuff, not atmos temperature
+  - type: Reflect
+    reflectProb: 0.75
```
Comment (keronshb) [inline Resources/Prototypes/Entities/Clothing/OuterClothing/armor.yml:98]:
```suggestion
    reflectProb: 1
```

Assuming there is only one per map in the armory, and if they can't be vended, ordered, crafted, or printed, bump this up to a 100% chance. Was like that in ss13. Yes it's extremely powerful which is why there's only one per map, and generally why it's a traitor objective.

File: Resources/Prototypes/Entities/Clothing/Uniforms/ship_vs_ship.yml [hunk 2]
```diff
@@ -93,4 +93,6 @@
   - type: Sprite
     sprite: Clothing/Uniforms/Jumpsuit/ce_syndie.rsi
   - type: Clothing
-    sprite: Clothing/Uniforms/Jumpsuit/ce_syndie.rsi
\ No newline at end of file
+    sprite: Clothing/Uniforms/Jumpsuit/ce_syndie.rsi
+  - type: Reflect
+    reflectProb: 0.5
```
Comment (keronshb) [inline Resources/Prototypes/Entities/Clothing/Uniforms/ship_vs_ship.yml:98]: Odd choice for making a syndie jumpsuit reflect. What's the reason for it?

File: Resources/Prototypes/Entities/Structures/Walls/walls.yml [hunk 3]
```diff
@@ -672,6 +672,8 @@
     damageModifierSet: Metallic
   - type: Physics
     bodyType: Static
+  - type: Reflect
+    reflectProb: 0.1
```
Comment (keronshb) [inline Resources/Prototypes/Entities/Structures/Walls/walls.yml:676]:
```suggestion
    reflectProb: 1
```

Shuttle walls should reflect 100% of the time.

File: Resources/Prototypes/Entities/Structures/Walls/walls.yml [hunk 4]
```diff
@@ -743,6 +745,8 @@
         acts: ["Destruction"]
     destroySound:
       path: /Audio/Effects/metalbreak.ogg
+  - type: Reflect
+    reflectProb: 0.1
```
Comment (keronshb) [inline Resources/Prototypes/Entities/Structures/Walls/walls.yml:749]:
```suggestion
    reflectProb: 1
```

Shuttle walls should reflect 100% of the time.

**Review by keronshb — CHANGES_REQUESTED** [review]
A few value tweaks for this review.

**Comment by graevy** [comment]
> A few value tweaks for this review.

if reflectprob is going to be 1 I need to maintain a uid:prob map for cumulative probability instead of a pure numeric solution because of div0 errors. I'll write it tn

File: Resources/Prototypes/Entities/Clothing/Uniforms/ship_vs_ship.yml (hunk 2 above)
Comment (graevy) [reply Resources/Prototypes/Entities/Clothing/Uniforms/ship_vs_ship.yml:98]: item desc says it has a reflective vest

File: Resources/Prototypes/Entities/Clothing/OuterClothing/armor.yml (hunk 1 above)
Comment (graevy) [reply Resources/Prototypes/Entities/Clothing/OuterClothing/armor.yml:98]: sure. i remember being able to order them with cargo egun crates on my old server so i went with 3/4

**Comment by graevy** [comment]
what other events cause item removal?

File: Resources/Prototypes/Entities/Clothing/Uniforms/ship_vs_ship.yml (hunk 2 above)
Comment (keronshb) [reply Resources/Prototypes/Entities/Clothing/Uniforms/ship_vs_ship.yml:98]: I think that means an emergency reflective vest a roadworker wears, not the armory one.

File: Resources/Prototypes/Entities/Clothing/OuterClothing/armor.yml (hunk 1 above)
Comment (keronshb) [reply Resources/Prototypes/Entities/Clothing/OuterClothing/armor.yml:98]: Just triple check there's only one per map and then you can make it 100%.

File: Resources/Prototypes/Entities/Clothing/Uniforms/ship_vs_ship.yml (hunk 2 above)
Comment (graevy) [reply Resources/Prototypes/Entities/Clothing/Uniforms/ship_vs_ship.yml:98]: my sleep-deprived ass completely missing this

**Comment by keronshb** [comment]
> what other events cause item removal?

What do you mean by item removal?

**Comment by graevy** [comment]
> > what other events cause item removal?
> 
> What do you mean by item removal?

items being removed from players

File: Content.Shared/Weapons/Reflect/SharedReflectSystem.cs [hunk 5]
```diff
@@ -160,4 +166,55 @@ private void OnReflectHitscan(EntityUid uid, ReflectComponent component, ref Hit
...
+            try
+            {
+                reflection.ReflectProb = (reflection.ReflectProb - comp.ReflectProb) / (1 - comp.ReflectProb);
+            }
+            catch (DivideByZeroException e)
+            {
+                reflection.ReflectProb = 0;
+                Logger.Error("component with reflectprob 1 should've been pruned", e.ToString());
+            }
+        }
```
Comment (metalgearsloth) [inline Content.Shared/Weapons/Reflect/SharedReflectSystem.cs:217]: Trycatch is for unexpected error handling and is also too slow for games, just handle it yourself.

**Comment by graevy** [comment]
okay, this is cleaner. the probabilities behaved as expected when tested. i hope i handled player slots correctly?

**Comment by keronshb** [comment]
> okay, this is cleaner. the probabilities behaved as expected when tested. i hope i handled player slots correctly? i used them instead of a mapping because it's safer. do you squash merge or should i rebase

We squash merge. Doesn't hurt to rebase tho cause that's how I do it.

You shouldn't be making PRs off of your master branch.

Otherwise this looks fine.

**Comment by graevy** [comment]
> Also please remove references to the Syndie CE suit since it's not part of the PR anymore.
> 
> Otherwise this looks fine.

done. i'll rebase my commits next time. thanks!

**Review by keronshb — APPROVED** [review]

**Comment by keronshb** [comment]
JUST noticed the CL.

Make the CL reflect the additions you made.

**Pull request merged**